component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/voiceotp/OneTimePasswordUtils.java

```diff
@@ -0,0 +1,247 @@
+     *                                  algorithms available.
+     * @throws InvalidKeyException      The secret provided was not a valid HMAC-SHA-256 key.
+     */
+
+    public static byte[] hmacShaGenerate(byte[] keyBytes, byte[] text)
+            throws NoSuchAlgorithmException, InvalidKeyException {
+
+        Mac hmacSha;
+        try {
+            hmacSha = Mac.getInstance(VoiceOTPConstants.ALGORITHM_HMAC);
+        } catch (NoSuchAlgorithmException nsa) {
+            hmacSha = Mac.getInstance(VoiceOTPConstants.ALGORITHM_HMAC_SHA);
+        }
+        SecretKeySpec macKey = new SecretKeySpec(keyBytes, "RAW");
+        hmacSha.init(macKey);
+        return hmacSha.doFinal(text);
+    }
+
+    /**
+     * This method generates an OTP value for the given set of parameters.
+     *
+     * @param secret           The shared secret
+     * @param movingFactor     The counter, or other value that changes on a per use
+     *                         basis.
+     * @param codeDigits       The number of digits in the OTP, not including the checksum,
+     *                         if any.
+     * @param addChecksum      A flag that indicates if a checksum digit
+     *                         should be appended to the OTP.
+     * @param truncationOffset The offset into the MAC result to begin truncation. If this
+     *                         value is out of the range of 0 ... 15, then dynamic truncation
+     *                         will be used. Dynamic truncation is when the last 4 bits of
+     *                         the last byte of the MAC are used to determine the start
+     *                         offset.
+     * @throws NoSuchAlgorithmException If no provider makes either HmacSHA256 or HMAC-SHA-256 digest
+     *                                  algorithms available.
+     * @throws InvalidKeyException      The secret provided was not a valid HMAC-SHA-256 key.
+     */
+    public static String generateOTP(byte[] secret, long movingFactor, int codeDigits, boolean addChecksum,
+                                     int truncationOffset) throws NoSuchAlgorithmException, InvalidKeyException {
+
+        // Put movingFactor value into text byte array.
+        String result = null;
+        int digits = addChecksum ? (codeDigits + 1) : codeDigits;
+        byte[] text = new byte[8];
+        for (int i = text.length - 1; i >= 0; i--) {
+            text[i] = (byte) (movingFactor & 0xff);
+            movingFactor >>= 8;
+        }
+        // Compute hmac hash.
+        byte[] hash = hmacShaGenerate(secret, text);
+
+        // Put selected bytes into result int.
+        int offset = hash[hash.length - 1] & 0xf;
+        if ((0 <= truncationOffset) && (truncationOffset < (hash.length - 4))) {
+            offset = truncationOffset;
+        }
+        int binary = ((hash[offset] & 0x7f) << 24) | ((hash[offset + 1] & 0xff) << 16)
+                | ((hash[offset + 2] & 0xff) << 8) | (hash[offset + 3] & 0xff);
```

Comment:
this code block is duplicated in `generateAlphaNumericOTP` as well. Can't we extract that into a private method and use within both the public methods

Reply:
There are few lines of code common in both methods. If we move to separate method , it would reduce the readability of the functions. It is better to keep it as it is to get an idea about the funtionality quickly. 

And the common funtionality is not a considarable logic to be separated.